drivers/Watchdog.cpp

```diff
@@ -18,48 +18,83 @@
 
 #include "Watchdog.h"
 
-namespace mbed {
-
-watchdog_status_t Watchdog::start(const uint32_t timeout)
+namespace mbed
 {
-    if (timeout == 0) {
-        return WATCHDOG_STATUS_INVALID_ARGUMENT;
-    }
 
-    if (timeout > max_timeout()) {
-        return WATCHDOG_STATUS_INVALID_ARGUMENT;
-    }
+Watchdog *Watchdog::first;
 
-    watchdog_config_t config;
-    config.timeout_ms = timeout;
+void Watchdog::add_to_list()
+{
+    this->next = first;
+    first = this;
+    _is_initialized =  true;
+}
 
-    return hal_watchdog_init(&config);
+void Watchdog::start()
+{
+    MBED_ASSERT(!_is_initialized);
+    core_util_critical_section_enter();
+    add_to_list();
+    core_util_critical_section_exit();
 }
 
 
 void Watchdog::kick()
 {
-    hal_watchdog_kick();
+    MBED_ASSERT(_is_initialized);
+    core_util_critical_section_enter();
+    _current_count = 0;
+    core_util_critical_section_exit();
 }
 
-
-watchdog_status_t Watchdog::stop()
+void Watchdog::stop()
 {
-    return hal_watchdog_stop();
+    MBED_ASSERT(_is_initialized);
+    core_util_critical_section_enter();
+    remove_from_list();
+    core_util_critical_section_exit();
 }
 
-
-uint32_t Watchdog::reload_value() const
+void Watchdog::remove_from_list()
 {
-    return hal_watchdog_get_reload_value();
+    Watchdog *cur_ptr = first,
+              *prev_ptr = NULL;
+    while(cur_ptr != NULL) {
+        if(cur_ptr == this) {
+            if(cur_ptr == first) {
+                prev_ptr = first;
+                first = cur_ptr->next;
+                prev_ptr->next = NULL;
+            } else {
+                prev_ptr->next = cur_ptr->next;
+                cur_ptr->next = NULL;
+            }
+            _is_initialized = false;
+            break;
+        } else {
+            prev_ptr = cur_ptr;
+            cur_ptr = cur_ptr->next;
+        }
+    }
 }
 
-
-uint32_t Watchdog::max_timeout()
+void Watchdog::process()
 {
-    const watchdog_features_t features = hal_watchdog_get_platform_features();
+    Watchdog *cur_ptr =  first;
+    while(cur_ptr != NULL) {
+        if(cur_ptr->_current_count > cur_ptr->_max_timeout) {
+            system_reset();
+        } else {
+            cur_ptr->_current_count += ((HW_WATCHDOG_TIMEOUT/2));
```

Comment:
`HW_WATCHDOG_TIMEOUT / 2` does not mean anything in this context, it should be passed as parameter instead

Reply:
Also should be changed to `(HW_WATCHDOG_TIMEOUT + 1) / 2` for the extreme case of `HW_WATCHDOG_TIMEOUT == 1`